lib/pbench/agent/tool_meister_client.py

```diff
@@ -41,195 +34,173 @@ class Client:
+
+    def _publish(self, group: str, directory: str, action: str, args=None) -> int:
+        """
+        A helper function that handles the publishing of messages via
+        state-signals, using the Client's SignalExporter. It then awaits
+        responses from subscribers (TDS).
 
         Returns 0 on success, 1 on failure; logs are also written for any
         errors encountered.
         """
-        if action not in api_tm_allowed_actions:
-            return 1
+
         # The published message contains four pieces of information:
         #   {
-        #     "action": "< 'start' | 'stop' | 'send' | 'kill' >",
+        #     "action": "< 'init' | 'start' | 'stop' | 'send' | 'end' | 'terminate' >",
         #     "group": "< the tool group name for the tools to operate on >",
         #     "directory": "< the local directory path to store collected data >"
         #     "args": "< arbitrary argument payload for a particular action >"
         #   }
         # The caller of tool-meister-client must be sure the directory argument
         # is accessible by the Tool Data Sink instance.
-        self.logger.debug("publish %s on chan %s", action, self.from_client_channel)
-        msg = dict(action=action, group=group, directory=str(directory), args=args)
+
+        # This check/conversion is necessary, as the pbench scripts sometimes pass
+        # in the directory as a str, and sometimes as pathlib.PosixPath
+        if directory and not isinstance(directory, str):
+            try:
+                directory = str(directory)
+            except Exception:
+                self.logger.exception(
+                    "Publish failed: directory must be compatible with type string, was instead %s",
+                    type(directory),
+                )
+        metadata = dict(group=group, directory=directory, args=args)
+        self.logger.debug(
+            "publish state signal for state %s with metadata: %s", action, metadata
+        )
         try:
-            num_present = self.redis_server.publish(
-                self.from_client_channel, json.dumps(msg)
+            resp, msgs = self.sig_pub.publish_signal(
+                event=action, tag="from_pbench_client", metadata=metadata, timeout=100
             )
         except Exception:
-            self.logger.exception("Failed to publish client message")
+            self.logger.exception("Failed to publish client signal")
             return 1
         else:
-            self.logger.debug("published %s", self.from_client_channel)
-            if num_present != 1:
-                self.logger.error(
-                    "Failed to publish to the TDS, encountered %d subscribers"
-                    " on the channel",
-                    num_present,
-                )
+            if resp != 0:
+                self.logger.error("Missing or bad response from the TDS, %s", str(resp))
                 ret_val = 1
+                for responder, msg in msgs.items():
```

Comment:
What does it mean to get more than one responder, msg pair?

Reply:
When publishing a signal, each responder can respond with an optional message (potentially including reasons why it failed or other desirable details). The user then receives a dict of all of the messages received in responses, indexed by the responders' IDs.

Reply:
We were not waiting for more than one responder before, were we?

Reply:
Nono, the most common use-case is still expected to be only one responder (a single TDS). It's just implemented like this for robustness/extensibility (if something goes wrong somewhere else, or there are multiple tool data sinks running, or if something new wants to be subscribing/responding to clients in the future, etc.). Just to make sure we are capturing everything that could be going on in the event of a failure or preparedness for future use cases.